scheduler/src/cook/mesos/task_stats.clj

```diff
@@ -0,0 +1,122 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.mesos.task-stats
+  (:require [clj-time.core :as t]
+            [cook.mesos.util :as util]
+            [datomic.api :as d]
+            [plumbing.core :as pc]))
+
+(defn- task->task-entry
+  "Updates the provided stats map with the data from the provided task entity"
+  [task-ent]
+  (let [job-ent (:job/_instance task-ent)
+        user (:job/user job-ent)
+        resources (util/job-ent->resources job-ent)
+        run-time (-> task-ent util/task-run-time)
+        seconds (-> run-time .toDuration .getMillis (/ 1000.0))]
+    (assoc resources
+      :user user
+      :run-time-seconds seconds
+      :cpu-seconds (* seconds (:cpus resources))
+      :mem-seconds (* seconds (:mem resources))
+      :reason (get-in task-ent [:instance/reason :reason/string]))))
+
+(defn- get-tasks
+  "Gets all tasks that started in the specified time range and with the specified status."
+  [db instance-status start end name-filter-fn]
+  (let [start-entity-id (d/entid-at db :db.part/user (.toDate (t/minus start (t/hours 1))))
+        end-entity-id (d/entid-at db :db.part/user (.toDate (t/plus end (t/hours 1))))
+        instance-status-entid (d/entid db instance-status)
+        instance-status-attribute-entid (d/entid db :instance/status)
+        start-millis (.getTime (.toDate start))
+        end-millis (.getTime (.toDate end))
+        tasks
+        (->> (d/seek-datoms db :avet :instance/status instance-status-entid start-entity-id)
```

Comment:
What is the performance of this query? Should the start-end range be restricted somehow?

Reply:
Here are some timings observed from a large internal database:

- failed instances for 1 day: ~60-85 seconds
- failed instances for 2 days: ~270 seconds
- failed instances for 5 days: ~430 seconds

I restricted the time range to a max of 31 days

Reply:
okay, should we be concerned about the impact of running this query on one of the query routers?

Reply:
I don't think we should be more concerned about this query than the query used by `/list`. This is also an admin-only endpoint.